internal/core/software_renderer.rs

```diff
@@ -1667,6 +1731,18 @@ impl<'a, T: ProcessScene> crate::item_rendering::ItemRenderer for SceneBuilder<'
                 self.draw_text_paragraph(paragraph, physical_clip, offset, color);
             }
         }
+
+        let cursor_rect = self
+            .window
+            .window_adapter()
+            .renderer()
+            .text_input_cursor_rect_for_byte_offset(
+                text_input,
+                text_input.cursor_position_byte_offset().round() as usize,
```

Comment:
This is wrong as well. The cursor is now drawn unconditionally, regardless of whether the `TextInput` wants the cursor to be shown or not. When you run the gallery example, you can see that the textedit's cursor is drawn even when the TextEdit is not visible. Call `visual_representation` on the `TextInput` and use the `cursor_position` field. If it's `None`, then you don't need to draw a cursor, and if it's `Some` then it holds the byte offset in the text to draw.

Regarding the text to draw: `draw_text_input` should also not draw ` let string = text.text();` but call `visual_representation()`, to take care of rendering password characters correctly.